File: src/handlers/handleClient.js
import { Buffer } from "node:buffer";
import { Socket } from "node:net";

import { POLICY_RESPONSE, characters } from "../config.js";
import {
    buildEnterGamePacket,
    buildGameInitPacket,
    buildHandshakeResponse,
    buildLoginChallenge,
    buildLoginCharacterListBitpacked,
} from "../build/exportBuilds.js";
import { BitReader } from "../classes/BitReader.js";
import { BitBuffer } from "../classes/BitBuffer.js";
import { buildEntityPacket } from "../build/entityPacket.js";

// Paket gönderme işlemini bir fonksiyona çıkaralım
function sendPacketSequence(socket, char) {
    return new Promise((resolve) => {
        // 1. Önce character acknowledgment
        const ackPkt = Buffer.alloc(4);
        ackPkt.writeUInt16BE(0x16, 0);
        ackPkt.writeUInt16BE(0, 2);
        socket.write(ackPkt);
        console.log("Sent character select acknowledgment (0x16)");

        // 2. Enter game packet
        setTimeout(() => {
            const enterPacket = buildEnterGamePacket();
            socket.write(enterPacket);
            console.log("Sent enter game packet (0x1A)");

            // 3. Game init packet
            setTimeout(() => {
                const initPkt = buildGameInitPacket();
                socket.write(initPkt);
                console.log("Sent game init packet (0x1B)");

                // 4. Entity/Paperdoll packet
                setTimeout(() => {
                    const entityXml = buildEntityPacket(char, "Player");
                    const buf = new BitBuffer();
                    buf.writeUtfString(entityXml);
                    const payload = buf.toBytes();

                    const pdPktHeader = Buffer.alloc(4);
                    pdPktHeader.writeUInt16BE(0x7c, 0);
                    pdPktHeader.writeUInt16BE(payload.length, 2);

                    const finalPacket = Buffer.concat([pdPktHeader, payload]);
                    socket.write(finalPacket);
                    console.log("Sent entity packet (0x7C)");

                    resolve();
                }, 300);
            }, 300);
        }, 300);
    });
}

function handleClient(socket = new Socket()) {
    console.log("Connection from", socket.remoteAddress);

    socket.on("error", (err) => {
        console.error("Socket error:", err);
    });

    socket.on("data", (data) => {
        if (data.includes(Buffer.from("<policy-file-request/>"))) {
            console.log("Flash policy request received. Sending policy XML.");
            socket.write(POLICY_RESPONSE);
            return;
        }

        const hexData = data.toString("hex");
        console.log("Received raw data:", hexData);

        if (hexData.length < 4) {
            return;
        }

        try {
            const pktType = parseInt(hexData.substring(0, 4), 16);

            if (pktType === 0x11) {
                const sessionId =
                    hexData.length >= 12
                        ? parseInt(hexData.substring(8, 12), 16)
                        : 0;
                console.log(
                    `Got handshake packet (0x11), session ID = ${sessionId}`
                );
                const resp = buildHandshakeResponse(sessionId);
                socket.write(resp);
                console.log(
                    "Sent handshake response (0x12):",
                    resp.toString("hex")
                );
                setTimeout(() => {
                    const challengePacket = buildLoginChallenge("CHALLENGE");
                    socket.write(challengePacket);
                    console.log(
                        "Sent login challenge (0x13):",
                        challengePacket.toString("hex")
                    );
                }, 200);
            } else if (pktType === 0x13 || pktType === 0x14) {
                console.log(
                    "Got authentication packet (0x13/0x14). Parsing..."
                );
                const pkt = buildLoginCharacterListBitpacked();
                socket.write(pkt);
                console.log(
                    "Sent login character list (0x15):",
                    pkt.toString("hex")
                );
            } else if (pktType === 0x17) {
                console.log(
                    "Got character creation packet (0x17). Parsing creation data..."
                );
                const payload = data.subarray(4);
                try {
                    const br = new BitReader(payload);
                    const name = br.readString();
                    const class_name = br.readString();
                    const computed = br.readString();
                    const extra1 = br.readString();
                    const extra2 = br.readString();
                    const extra3 = br.readString();
                    const extra4 = br.readString();
                    const hair_color = br.readBits(24);
                    const skin_color = br.readBits(24);
                    const shirt_color = br.readBits(24);
                    const pant_color = br.readBits(24);

                    // Set default gear based on class
                    let default_gear;
                    if (class_name.toLowerCase() === "mage") {
                        default_gear = `
                            <Item Slot='MainHand' ID='1002' Scale='1.0'/>
                            <Item Slot='OffHand' ID='1003' Scale='0.8'/>
                        `;
                    } else if (class_name.toLowerCase() === "rogue") {
                        default_gear = `
                            <Item Slot='Offhand' ID='${ROGUE_ITEMS["Offhand"]["ID"]}' Scale='${ROGUE_ITEMS["Offhand"]["Scale"]}'/>
                            <Item Slot='WholeOffhand' ID='${ROGUE_ITEMS["WholeOffhand"]["ID"]}' Scale='${ROGUE_ITEMS["WholeOffhand"]["Scale"]}'/>
                        `;
                    } else if (class_name.toLowerCase() === "paladin") {
                        default_gear = `
                            <Item Slot='MainHand' ID='${PALADIN_ITEMS["MainHand"]["ID"]}' Scale='${PALADIN_ITEMS["MainHand"]["Scale"]}'/>
                        `;
                    } else {
                        // Mage and other classes
                        default_gear =
                            "<Item Slot='1' ID='1001' Name='StarterSword' Scale='1.0'/>";
                    }

                    // Create new character
                    const new_char = [
                        name,
                        class_name,
                        1,
                        computed,
                        extra1,
                        extra2,
                        extra3,
                        extra4,
                        hair_color,
                        skin_color,
                        shirt_color,
                        pant_color,
                        default_gear,
                    ];

                    // Debug info
                    console.log("Parsed Character Creation Packet:");
                    console.log("  Name:     ", name);
                    console.log("  ClassName:", class_name);
                    console.log("  Extra:    ", [
                        computed,
                        extra1,
                        extra2,
                        extra3,
                        extra4,
                    ]);
                    console.log("  Colors:   ", [
                        hair_color,
                        skin_color,
                        shirt_color,
                        pant_color,
                    ]);
                    console.log("  Gear:     ", default_gear);

                    // Add character once
                    characters.push(new_char);
                    console.log(
                        `Created new char: userID=1, name='${name}', class='${class_name}'`
                    );

                    const pkt = buildLoginCharacterListBitpacked();
                    socket.write(pkt);
                    console.log(
                        "Sent updated login character list (0x15):",
                        pkt.toString("hex")
                    );

                    setTimeout(() => {
                        const ackPkt = Buffer.alloc(4);
                        ackPkt.writeUInt16BE(0x16, 0);
                        ackPkt.writeUInt16BE(0, 2);
                        socket.write(ackPkt);
                        console.log(
                            "Sent character select acknowledgment (0x16):",
                            ackPkt.toString("hex")
                        );

                        setTimeout(() => {
                            const enterPacket = buildEnterGamePacket();
                            socket.write(enterPacket);
                            console.log(
                                "Sent enter game packet (0x1A):",
                                enterPacket.toString("hex")
                            );

                            setTimeout(() => {
                                const initPkt = buildGameInitPacket();
                                socket.write(initPkt);
                                console.log(
                                    "Sent game init packet (0x1B):",
                                    initPkt.toString("hex")
                                );

                                setTimeout(() => {
                                    const paperdollXml = buildEntityPacket(
                                        new_char,
                                        "CharCreateUI"
                                    );
                                    const buf = new BitBuffer();
                                    buf.writeUtfString(paperdollXml);
                                    const pdPayload = buf.toBytes();
                                    const pdPktHeader = Buffer.alloc(4);
                                    pdPktHeader.writeUInt16BE(0x7c, 0);
                                    pdPktHeader.writeUInt16BE(
                                        pdPayload.length,
                                        2
                                    );
                                    const pdPkt = Buffer.concat([
                                        pdPktHeader,
                                        pdPayload,
                                    ]);
                                    socket.write(pdPkt);
                                    console.log(
                                        "Sent paperdoll update (0x7C):",
                                        pdPkt.toString("hex")
                                    );
                                }, 200);
                            }, 200);
                        }, 200);
                    }, 200);
                } catch (e) {
                    console.log("Error parsing create character packet:", e);
                }
            } else if (pktType === 0x16) {
                console.log("Got character select packet (0x16)");

                const br = new BitReader(data.subarray(4));
                const selectedName = br.readString();

                // Karakteri bul
                const selectedChar = characters.find(
                    (char) => char[0] === selectedName
                );

                if (selectedChar) {
                    try {
                        sendPacketSequence(socket, selectedChar).catch(
                            (err) => {
                                console.error(
                                    "Error sending packet sequence:",
                                    err
                                );
                            }
                        );
                    } catch (err) {
                        console.error("Error sending packet sequence:", err);
                    }
                } else {
                    console.log(`Character '${selectedName}' not found`);
                }
            } else if (pktType === 0x19) {
                console.log(
                    "Got packet type 0x19. Request for character details."
                );
                const payload = data.subarray(4); // Skip 4-byte header (type + length)
                try {
                    const br = new BitReader(payload);
                    const name = br.readString();
                    console.log(`Requested character: ${name}`);

                    // Find character by name
                    let found = false;
                    for (const char of characters) {
                        if (char[0] === name) {
                            const xml = buildEntityPacket(char, "Player");
                            const buf = new BitBuffer();
                            buf.writeUtfString(xml);
                            const pdPayload = buf.toBytes();
                            const pdPktHeader = Buffer.alloc(4);
                            pdPktHeader.writeUInt16BE(0x7c, 0);
                            pdPktHeader.writeUInt16BE(pdPayload.length, 2);
                            const pdPkt = Buffer.concat([
                                pdPktHeader,
                                pdPayload,
                            ]);
                            socket.write(pdPkt);
                            console.log(
                                "Sent paperdoll update (0x7C):",
                                pdPkt.toString("hex")
                            );
                            found = true;
                            break;
                        }
                    }

                    if (!found) {
                        console.log(`Character '${name}' not found.`);
                        const ackPkt = Buffer.alloc(4);
                        ackPkt.writeUInt16BE(0x19, 0);
                        ackPkt.writeUInt16BE(0, 2);
                        socket.write(ackPkt);
                        console.log("Sent 0x19 ack:", ackPkt.toString("hex"));
                    }
                } catch (e) {
                    console.log("Error parsing 0x19 packet:", e);
                    const ackPkt = Buffer.alloc(4);
                    ackPkt.writeUInt16BE(0x19, 0);
                    ackPkt.writeUInt16BE(0, 2);
                    socket.write(ackPkt);
                    console.log("Sent 0x19 ack:", ackPkt.toString("hex"));
                }
            } else if (pktType === 0x7c) {
                console.log(
                    "Received packet type 0x7C. (Appearance/cue update)"
                );
                if (characters.length > 0) {
                    const entityXml = buildEntityPacket(
                        characters[0],
                        "Player"
                    );
                    const buf = new BitBuffer();
                    buf.writeUtfString(entityXml);
                    const payload = buf.toBytes();
                    const responseHeader = Buffer.alloc(4);
                    responseHeader.writeUInt16BE(0x7c, 0);
                    responseHeader.writeUInt16BE(payload.length, 2);
                    const response = Buffer.concat([responseHeader, payload]);
                    socket.write(response);
                    console.log(
                        "Sent entity packet (0x7C):",
                        response.toString("hex")
                    );
                } else {
                    console.log(
                        "No character data available. Sending empty 0x7C response."
                    );
                    const response = Buffer.alloc(4);
                    response.writeUInt16BE(0x7c, 0);
                    response.writeUInt16BE(0, 2);
                    socket.write(response);
                    console.log(
                        "Sent 0x7C response:",
                        response.toString("hex")
                    );
                }
            }
        } catch (e) {
            console.log("Error processing packet:", e);
        }
    });

    socket.on("close", () => {
        console.log("Client disconnected.");
    });
}

export { handleClient };

File: src/handlers/handle_client.py
import time
import struct
from ..config import POLICY_RESPONSE, characters, ROGUE_ITEMS, PALADIN_ITEMS
from ..classes.bit_reader import BitReader
from ..classes.bit_buffer import BitBuffer
from ..build.export_builds import (
    build_handshake_response,
    build_enter_game_packet,
    build_game_init_packet,
    build_login_character_list_bitpacked,
    build_login_challenge,
    build_paperdoll_packet
)
from ..build.entity_packet import build_entity_packet

def handle_client(conn, addr):
    print("Connection from", addr)
    try:
        while True:
            data = conn.recv(4096)
            if not data:
                break

            if b"<policy-file-request/>" in data:
                print("Flash policy request received. Sending policy XML.")
                conn.sendall(POLICY_RESPONSE)
                continue

            hex_data = data.hex()
            print("Received raw data:", hex_data)

            if len(hex_data) < 4:
                continue

            try:
                pkt_type = int(hex_data[:4], 16)
            except ValueError:
                print("Error parsing packet type.")
                continue

            if pkt_type == 0x11:
                session_id = int(hex_data[8:12], 16) if len(hex_data) >= 12 else 0
                print(f"Got handshake packet (0x11), session ID = {session_id}")
                resp = build_handshake_response(session_id)
                conn.sendall(resp)
                print("Sent handshake response (0x12):", resp.hex())
                time.sleep(0.2)
                challenge_packet = build_login_challenge("CHALLENGE")
                conn.sendall(challenge_packet)
                print("Sent login challenge (0x13):", challenge_packet.hex())
                time.sleep(0.2)

            elif pkt_type in (0x13, 0x14):
                print("Got authentication packet (0x13/0x14). Parsing...")
                pkt = build_login_character_list_bitpacked()
                conn.sendall(pkt)
                print("Sent login character list (0x15):", pkt.hex())
                time.sleep(0.2)

            elif pkt_type == 0x17:
                print("Got character creation packet (0x17). Parsing creation data...")
                payload = data[4:]
                try:
                    br = BitReader(payload)
                    name = br.read_string()
                    class_name = br.read_string()
                    computed = br.read_string()
                    extra1 = br.read_string()
                    extra2 = br.read_string()
                    extra3 = br.read_string()
                    extra4 = br.read_string()
                    hair_color = br.read_bits(24)
                    skin_color = br.read_bits(24)
                    shirt_color = br.read_bits(24)
                    pant_color = br.read_bits(24)

                    # Sınıfa göre varsayılan donanımı ayarla
                    if class_name.lower() == "mage":
                        default_gear = """
                            <Item Slot='MainHand' ID='1002' Scale='1.0'/>
                            <Item Slot='OffHand' ID='1003' Scale='0.8'/>
                        """
                    elif class_name.lower() == "rogue":
                        default_gear = f"""
                            <Item Slot='Offhand' ID='{ROGUE_ITEMS["Offhand"]["ID"]}' Scale='{ROGUE_ITEMS["Offhand"]["Scale"]}'/>
                            <Item Slot='WholeOffhand' ID='{ROGUE_ITEMS["WholeOffhand"]["ID"]}' Scale='{ROGUE_ITEMS["WholeOffhand"]["Scale"]}'/>
                        """
                    elif class_name.lower() == "paladin":
                        default_gear = f"""
                            <Item Slot='MainHand' ID='{PALADIN_ITEMS["MainHand"]["ID"]}' Scale='{PALADIN_ITEMS["MainHand"]["Scale"]}'/>
                        """
                    else:  # Mage ve diğer sınıflar için
                        default_gear = "<Item Slot='1' ID='1001' Name='StarterSword' Scale='1.0'/>"

                    # Yeni karakter oluştur
                    new_char = (
                        name, class_name, 1, computed, extra1, extra2, extra3, extra4,
                        hair_color, skin_color, shirt_color, pant_color, default_gear
                    )
                    
                    # Debug bilgisi
                    print("Parsed Character Creation Packet:")
                    print("  Name:     ", name)
                    print("  ClassName:", class_name)
                    print("  Extra:    ", [computed, extra1, extra2, extra3, extra4])
                    print("  Colors:   ", [hair_color, skin_color, shirt_color, pant_color])
                    print("  Gear:     ", default_gear)

                    # Karakteri bir kez ekle
                    characters.append(new_char)
                    print(f"Created new char: userID=1, name='{name}', class='{class_name}'")

                except Exception as e:
                    print("Error parsing create character packet:", e)
                    continue

                pkt = build_login_character_list_bitpacked()
                conn.sendall(pkt)
                print("Sent updated login character list (0x15):", pkt.hex())
                time.sleep(0.2)

                ack_pkt = struct.pack(">HH", 0x16, 0)
                conn.sendall(ack_pkt)
                print("Sent character select acknowledgment (0x16):", ack_pkt.hex())
                time.sleep(0.2)

                enter_packet = build_enter_game_packet()
                conn.sendall(enter_packet)
                print("Sent enter game packet (0x1A):", enter_packet.hex())
                time.sleep(0.2)

                init_pkt = build_game_init_packet()
                conn.sendall(init_pkt)
                print("Sent game init packet (0x1B):", init_pkt.hex())
                time.sleep(0.2)

                paperdoll_xml = build_entity_packet(new_char, category="CharCreateUI")
                buf = BitBuffer()
                buf.write_utf_string(paperdoll_xml)
                pd_payload = buf.to_bytes()
                pd_pkt = struct.pack(">HH", 0x7C, len(pd_payload)) + pd_payload
                conn.sendall(pd_pkt)
                print("Sent paperdoll update (0x7C):", pd_pkt.hex())
                time.sleep(0.2)

            elif pkt_type == 0x16:
                print("Got character select packet (0x16).")
                
                # Önce karakter bilgilerini gönder
                for char in characters:
                    br = BitReader(data[4:])  # Skip the header
                    selected_name = br.read_string()
                    if char[0] == selected_name:
                        paperdoll_xml = build_entity_packet(char, category="Player")
                        buf = BitBuffer()
                        buf.write_utf_string(paperdoll_xml)
                        pd_payload = buf.to_bytes()
                        pd_pkt = struct.pack(">HH", 0x7C, len(pd_payload)) + pd_payload
                        conn.sendall(pd_pkt)
                        print("Sent paperdoll update (0x7C)")
                        time.sleep(0.1)
                        break
                
                # Sonra diğer paketleri gönder
                ack_pkt = struct.pack(">HH", 0x16, 0)
                conn.sendall(ack_pkt)
                time.sleep(0.1)
                
                enter_packet = build_enter_game_packet()
                conn.sendall(enter_packet)
                time.sleep(0.1)
                
                init_pkt = build_game_init_packet()
                conn.sendall(init_pkt)
                time.sleep(0.1)

            elif pkt_type == 0x19:
                print("Got packet type 0x19. Request for character details.")
                payload = data[4:]  # Skip 4-byte header (type + length)
                br = BitReader(payload)
                try:
                    name = br.read_string()
                    print(f"Requested character: {name}")
                    # Find character by name
                    for char in characters:
                        if char[0] == name:
                            xml = build_entity_packet(char, category="Player")
                            buf = BitBuffer()
                            buf.write_utf_string(xml)
                            pd_payload = buf.to_bytes()
                            pd_pkt = struct.pack(">HH", 0x7C, len(pd_payload)) + pd_payload
                            conn.sendall(pd_pkt)
                            print("Sent paperdoll update (0x7C):", pd_pkt.hex())
                            break
                    else:
                        print(f"Character '{name}' not found.")
                        ack_pkt = struct.pack(">HH", 0x19, 0)
                        conn.sendall(ack_pkt)
                        print("Sent 0x19 ack:", ack_pkt.hex())
                except Exception as e:
                    print("Error parsing 0x19 packet:", e)
                    ack_pkt = struct.pack(">HH", 0x19, 0)
                    conn.sendall(ack_pkt)
                    print("Sent 0x19 ack:", ack_pkt.hex())
                time.sleep(0.2)

            elif pkt_type == 0x7C:
                print("Received packet type 0x7C. (Appearance/cue update)")
                if characters:
                    entity_xml = build_entity_packet(characters[0], category="Player")
                    buf = BitBuffer()
                    buf.write_utf_string(entity_xml)
                    payload = buf.to_bytes()
                    response = struct.pack(">HH", 0x7C, len(payload)) + payload
                    conn.sendall(response)
                    print("Sent entity packet (0x7C):", response.hex())
                else:
                    print("No character data available. Sending empty 0x7C response.")
                    response = struct.pack(">HH", 0x7C, 0)
                    conn.sendall(response)
                    print("Sent 0x7C response:", response.hex())
                time.sleep(0.2)

    except Exception as e:
        print("Error:", e)
    finally:
        conn.close()
        print("Client disconnected.")
